modules/platforms/cpp/client/include/ignite/ignite_client.h

```diff
@@ -0,0 +1,131 @@
+#include <memory>
+
+#include "common/Config.h"
+#include "common/ignite_result.h"
+
+#include "ignite/ignite_client_configuration.h"
+#include "ignite/table/tables.h"
+
+namespace ignite
+{
+
+namespace detail
+{
+class IgniteClientImpl;
+}
+
+/**
+ * Ignite client.
+ */
+class IgniteClient
+{
+public:
+    // Deleted
+    IgniteClient(const IgniteClient&) = delete;
+    IgniteClient& operator=(const IgniteClient&) = delete;
+
+    // Default
+    IgniteClient() = default;
+    ~IgniteClient() = default;
+    IgniteClient(IgniteClient&&) = default;
+    IgniteClient& operator=(IgniteClient&&) = default;
+
+    /**
+     * Start client asynchronously.
+     *
+     * Client tries to establish connection to every endpoint. First endpoint is
+     * selected randomly. After that round-robin is used to determine the next
+     * address to establish connection to.
+     *
+     * System means are used to resolve endpoint IP addresses. If more than one
+     * IP address is returned, client attempts to connect to them in random order.
+     *
+     * Only one connection establishment can be in process at the same time.
+     *
+     * Client considered connected to a cluster when there is at least one
+     * connection to any node of the cluster. Upon this event, future will be set
+     * with a usable IgniteClient instance.
+     *
+     * @param configuration Client configuration.
+     * @param timeout Operation timeout.
+     * @param callback Callback to be called once operation is complete.
+     */
+    IGNITE_API static void startAsync(IgniteClientConfiguration configuration, std::chrono::milliseconds timeout,
+        IgniteCallback<IgniteClient> callback);
+
+    /**
+     * Start client synchronously.
+     *
+     * @see startAsync for details.
+     *
+     * @param configuration Client configuration.
+     * @param timeout Operation timeout.
+     * @return IgniteClient instance.
+     */
+    IGNITE_API static IgniteClient start(IgniteClientConfiguration configuration, std::chrono::milliseconds timeout);
+
+    /**
+     * Get client configuration.
+     *
+     * @return Configuration.
+     */
+    [[nodiscard]]
+    IGNITE_API const IgniteClientConfiguration& getConfiguration() const noexcept;
+
+    /**
+     * Get the table API.
+     *
+     * @return Table API.
+     */
+    [[nodiscard]]
+    IGNITE_API Tables getTables() const noexcept;
+
+private:
+    /**
+     * Constructor
+     *
+     * @param impl Implementation
+     */
+    explicit IgniteClient(std::shared_ptr<void> impl);
+
+    /**
+     * Get implementation reference.
+     *
+     * @return Implementation reference.
+     */
+    [[nodiscard]]
+    detail::IgniteClientImpl& getImpl() noexcept;
+
+    /**
+     * Get implementation reference.
+     *
+     * @return Implementation reference.
+     */
+    [[nodiscard]]
+    const detail::IgniteClientImpl& getImpl() const noexcept;
+
+    /** Implementation. */
+    std::shared_ptr<void> m_impl;
```

Comment:
I considered this but this is more error prone solution, which saves us nothing, because `IgniteClient` is never copied and when it's started there a lot of much more heavy operations.

Reply:
At least it saves some memory. And I don't see how unqiue_ptr is more error prone. But this is up to you.